include/dlaf/communication/kernels/all_reduce.h

```diff
@@ -89,52 +72,60 @@ void scheduleAllReduce(CommSender&& pcomm, MPI_Op reduce_op,
+  using dlaf::internal::withTemporaryTile;
+
+  auto all_reduce_final = [reduce_op, pcomm = std::forward<CommSender>(pcomm),
+                           tile_in =
+                               std::forward<TileInSender>(tile_in)](auto const& tile_out_comm) mutable {
+    auto all_reduce = [reduce_op, pcomm = std::move(pcomm),
+                       &tile_out_comm](auto const& tile_in_comm) mutable {
+      return whenAllLift(std::move(pcomm), reduce_op, std::cref(tile_in_comm),
+                         std::cref(tile_out_comm)) |
+             transformMPI(allReduce_o);
+    };
+    return withTemporaryTile<Device::CPU, CopyToDestination::Yes, CopyFromDestination::No,
+                             RequireContiguous::Yes>(std::move(tile_in), std::move(all_reduce));
+  };
+
+  // TODO: Can reductions happen on CPU only?
+  return withTemporaryTile<Device::CPU, CopyToDestination::No, CopyFromDestination::Yes,
+                           RequireContiguous::Yes>(std::forward<TileOutSender>(tile_out),
+                                                   std::move(all_reduce_final));
+}
 
+template <class CommSender, class TileSender>
+[[nodiscard]] auto scheduleAllReduceInPlace(CommSender&& pcomm, MPI_Op reduce_op, TileSender&& tile) {
   // Note:
   //
   // TILE ---> makeContiguous --+--> CONT_BUF ----> mpi_call ---> CONT_BUF --+
   //                            |                                            |
   //                            +------------------> TILE -------------------+-> copyBack ---> TILE
   //
-  //
   // The last TILE after the copyBack is returned so that other task can be attached to it,
   // AFTER the asynchronous MPI_AllReduce has completed
-  return std::forward<TSender>(tile) | ex::transfer(getBackendScheduler<Backend::MC>()) |
-         ex::let_value([pcomm = std::forward<CommSender>(pcomm),
-                        reduce_op](matrix::Tile<T, Device::CPU>& tile) mutable {
-           auto tile_reduced = whenAllLift(std::move(pcomm), reduce_op, makeItContiguous(tile)) |
-                               transformMPI(internal::allReduceInPlace_o);
-           return whenAllLift(std::move(tile_reduced), std::cref(tile)) |
-                  transform(Policy<Backend::MC>(thread_priority::high), copyBack_o) |
-                  ex::then([&tile]() { return std::move(tile); });
-         });
-}
+  using dlaf::comm::internal::allReduceInPlace_o;
+  using dlaf::comm::internal::transformMPI;
+  using dlaf::internal::CopyFromDestination;
+  using dlaf::internal::CopyToDestination;
+  using dlaf::internal::RequireContiguous;
+  using dlaf::internal::whenAllLift;
+  using dlaf::internal::withTemporaryTile;
+
+  auto all_reduce_in_place = [reduce_op,
+                              pcomm = std::forward<CommSender>(pcomm)](auto const& tile_comm) mutable {
+    return whenAllLift(std::move(pcomm), reduce_op, std::cref(tile_comm)) |
+           transformMPI(allReduceInPlace_o);
+  };
+
+  // TODO: Can reductions happen on CPU only?
```

Comment:
@rasolca do you happen to know if reductions could be made on GPU buffers if CUDA RDMA is enabled? I think to be on the safe side we should probably stay with explicitly using `Device::CPU` here until we get to test CUDA RDMA with a working implementation, but I'm wondering if I should update this comment in some way.

Reply:
They work with GPU pointers too, however the implementation I checked was reducing on CPU and copying back to GPU.

Reply:
Thanks. In that case it sounds like I can add to the general comment for the reduction functions that we explicitly choose to do the reduction on `Device::CPU` so that we can manage the copies, instead of having MPI do it for us. Sounds good?

Reply:
Updated the comment.